src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs

```diff
@@ -0,0 +1,180 @@
+            {
+                MarkEntityProcessed(symbol, _pendingSymbols);
+            }
+
+            public bool TryStartAnalyzingDeclaration(SyntaxReference decl, out DeclarationAnalyzerStateData state)
+            {
+                return TryStartProcessingEntity(decl.GetSyntax(), _pendingDeclarations, out state);
+            }
+
+            public void MarkDeclarationComplete(SyntaxReference decl)
+            {
+                MarkEntityProcessed(decl.GetSyntax(), _pendingDeclarations);
+            }
+
+            public bool TryStartSyntaxAnalysis(SyntaxTree tree, out AnalyzerStateData state)
+            {
+                return TryStartProcessingEntity(tree, _pendingSyntaxAnalysisTrees, out state);
+            }
+
+            public void MarkSyntaxAnalysisComplete(SyntaxTree tree)
+            {
+                MarkEntityProcessed(tree, _pendingSyntaxAnalysisTrees);
+            }
+
+            public void MarkDeclarationsComplete(ISymbol symbol)
+            {
+                foreach (var syntaxRef in symbol.DeclaringSyntaxReferences)
+                {
+                    MarkEntityProcessed(syntaxRef.GetSyntax(), _pendingDeclarations);
+                }
+            }
+
+            public void OnCompilationEventGenerated(CompilationEvent compilationEvent)
+            {
+                var symbolEvent = compilationEvent as SymbolDeclaredCompilationEvent;
+                if (symbolEvent != null)
+                {
+                    var needsAnalysis = false;
+                    var symbol = symbolEvent.Symbol;
+                    if (!AnalysisScope.ShouldSkipSymbolAnalysis(symbol))
+                    {
+                        needsAnalysis = true;
+                        _pendingSymbols.TryAdd(symbol, null);
+                    }
+
+                    if (!AnalysisScope.ShouldSkipDeclarationAnalysis(symbol))
+                    {
+                        foreach (var syntaxRef in symbol.DeclaringSyntaxReferences)
+                        {
+                            needsAnalysis = true;
+                            _pendingDeclarations.TryAdd(syntaxRef.GetSyntax(), null);
+                        }
+                    }
+
+                    if (!needsAnalysis)
+                    {
+                        return;
+                    }
+                }
+
+                _pendingEvents.TryAdd(compilationEvent, null);
+            }
+
+            public bool IsEventAnalyzed(CompilationEvent compilationEvent)
+            {
+                return IsEntityFullyProcessed(compilationEvent, _pendingEvents);
+            }
+
+            public void AnalyzeSymbolEvent(SymbolDeclaredCompilationEvent symbolDeclaredEvent)
+            {
+                if (IsEntityFullyProcessed(symbolDeclaredEvent, _pendingEvents))
+                {
+                    return;
+                }
+
+                if (!IsEntityFullyProcessed(symbolDeclaredEvent.Symbol, _pendingSymbols))
```

Comment:
I think this method needs a comment at the top. It is basically invoked after a specific event finishes getting processed (or cancelled) for any given diagnostics request. It tries to check upfront if the event is already marked as processed (say some other thread already finished analyzing this event, but this thread still happened to get the event in the queue). If not, then it checks if all of the sub-entities (symbol and all of its declarations) are fully processed (all its symbol actions, node actions and code block actions are executed). If so, then it marks the event as symbol declared event as fully processed.